Orange/canvas/application/addons.py

```diff
@@ -760,127 +770,170 @@ def setStatusMessage(self, message):
+                if self.conda:
+                    self.conda.upgrade(pkg.installable)
+                self.pip.upgrade(pkg.installable)
+            elif command == Uninstall:
+                self.setStatusMessage(
+                    "Uninstalling {}".format(pkg.local.project_name))
+                if self.conda:
+                    self.conda.uninstall(pkg.local)
+                self.pip.uninstall(pkg.local)
+        except CommandFailed as ex:
+            self.error.emit(
+                "Command failed: python {}".format(ex.cmd),
+                pkg, ex.retcode, ex.output
+            )
+            return
 
         if self.__queue:
             QTimer.singleShot(0, self._next)
         else:
             self.finished.emit()
 
-    def __subprocessrun(self, process):
-        output = []
-        while process.poll() is None:
-            try:
-                line = process.stdout.readline()
-            except IOError as ex:
-                if ex.errno != errno.EINTR:
-                    raise
-            else:
-                output.append(line)
-                print(line, end="")
-        # Read remaining output if any
-        line = process.stdout.read()
-        if line:
+
+class PipInstaller:
+    def __init__(self, user_install=False):
+        self.user_install = user_install
+
+    def install(self, pkg):
+        cmd = ["python", "-m", "pip", "install"]
+        if self.user_install:
+            cmd.append("--user")
+        cmd.append(pkg.name)
+
+        run_command(cmd)
+
+    def upgrade(self, package):
+        # This is done in two steps to avoid upgrading
+        # all of the dependencies - faster
+        self.upgrade_no_deps(package)
+        self.install(package)
+
+    def upgrade_no_deps(self, package):
+        cmd = ["python", "-m", "pip", "install", "--upgrade", "--no-deps"]
+        if self.user_install:
+            cmd.append("--user")
+        cmd.append(package.name)
+
+        run_command(cmd)
+
+    def uninstall(self, dist):
+        cmd = ["python", "-m", "pip", "uninstall", "--yes", dist.project_name]
+        run_command(cmd)
+
+        if self.user_install:
+            # Remove the package forcefully; pip doesn't (yet) uninstall
+            # --user packages (or any package outside sys.prefix?)-
+            # google: pip "Not uninstalling ?" "outside environment"
+            self.manual_uninstall(dist)
+
+    def manual_uninstall(self, dist):
+        install_path = os.path.join(
+            USER_SITE, re.sub('[^\w]', '_', dist.project_name))
+        pip_record = next(iglob(install_path + '*.dist-info/RECORD'),
+                          None)
+        if pip_record:
+            with open(pip_record) as f:
+                files = [line.rsplit(',', 2)[0] for line in f]
+        else:
+            files = [os.path.join(
+                USER_SITE, 'orangecontrib',
+                dist.project_name.split('-')[-1].lower()), ]
```

Comment:
This assumes all user-installed add-ons are in `orangecontrib` package, but this is just an internal convention, with the real add-on being anything that provides an `orange3.addon` entry point.

Reply:
So I should wrap this in try: except: pass?

This is the same code we are (using at the moment)[https://github.com/biolab/orange3/blame/master/Orange/canvas/application/addons.py#L844], I have just moved it to a separate class.

Reply:
I know, and I wrote it, but it's wrong in its assumptions.

No idea how to make it correct.

HTH